### PR TITLE
[905] Add Multiple Selection for Select with Optgroup

### LINKED ISSUE #905: Twig helpers for select don't allow pre selected values with multiple
Can't use `selected` option with `multiple` option `true` in the twig helpers for `select` and `select2`.

Additionally, this stops being able to add a `selected` option to the `choice` twig helper.

## Test case

```twig
{{
    form.select({
        'label': 'Deselect some',
        'options': [
            { 'label': 'a', 'value': 'a' },
            { 'label': 'b', 'value': 'b' },
        ],
        'selected': ['a', 'b'],
    })
}}
```

## Expected

A select is rendered with both `a` and `b` options pre-selected.

## Actual

Select is loaded with no selected options.

### PR DESCRIPTION
**Pull request by nvasileiadis**
Adds the ability to select multiple `option` elements in select/select2 helpers.

**Example:**
```
{{
    form.select({
        'selected': ['foo', 'bar'],
        'options': [
            {
                'label': 'bar',
                'value': 'foo'
            },
            {
                'label': 'baz',
                'value': 'bar'
            }
        ]
    })
}}
```

Closes #905 

## Breaking change checklist

The following points should be considered as an early-warning of introducing a breaking change to a Continuum platform product. This is not an exhaustive list of what constitutes a breaking change.

| Does this pull request... | Yes / No |
| ---- | ---- |
| require changes to `main.js` | ✅ No |
| require changes to `index.js` | ✅ No |
| require changes to `pulsar.scss` | ✅ No |
| require changes to `gruntfile.js` | ✅ No |
| require changes to `package.json` (not including dev dependencies) | ✅ No |
| require changes to `base.html.twig` (or other core views like header/footer) | ✅ No |
| require new arguments to be passed to a js component | ✅ No |
| require markup changes to maintain functionality | ✅ No |
| require changes to existing unit tests | ✅ No |

### CODE REVIEW

**Comment by Stanton** [comment]
@jackbentley Pushed changes which correctly implement this behaviour, test cases added also. I've sent you an collaborate invitation for the Pulsar project so I can request your review of this change.

File: views/pulsar/v2/helpers/elem.html.twig [hunk 1]
```diff
@@ -260,7 +260,7 @@ value       | string | Specifies the value of the input
                     {% if option is iterable %}
                         {% for item in option %}
                             <option {{ util.value(item.value) }}
-                                {%- if mainOptions.selected is defined and item.value == mainOptions.selected %} selected{% endif -%}
+                                {%- if mainOptions.selected is defined and item.value in mainOptions.selected %} selected{% endif -%}
```
Comment (mikedevelops) [inline views/pulsar/v2/helpers/elem.html.twig]: If we're changing the selected option to accept both strings and arrays, will `in` partially match strings here and potentially select values we don't want?

So `'foo' in ['foo', 'bar'] === 'foo'` whereas `'foo' in 'foobar' === 'foo'` ?

Comment (Stanton) [reply views/pulsar/v2/helpers/elem.html.twig]: The docs do make it sound that way, but it didn't work in my tests.

**Given** 

```
{{
    form.select({
        'selected': ['foo', 'ba'],
        'options': [
            {
                'label': 'bar',
                'value': 'foo'
            },
            {
                'label': 'baz',
                'value': 'bar'
            }
        ]
    })
}}
```

**Then**

```
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'<div class="form__group"><div class="controls"><select class="form__control select"><option value="foo" selected>bar</option><option value="bar" selected>baz</option></select></div></div>'
+'<div class="form__group"><div class="controls"><select class="form__control select"><option value="foo" selected>bar</option><option value="bar">baz</option></select></div></div>'
```

Comment (Stanton) [reply views/pulsar/v2/helpers/elem.html.twig]: I'll add test cases to test this behaviour, in case the twig internals change at some point

Comment (mikedevelops) [reply views/pulsar/v2/helpers/elem.html.twig]: your test there passes an array in the selected option, the issue should rear it's head if you changed your test to:

```{{
    form.select({
        'selected': 'ba',
        'options': [
            {
                'label': 'bar',
                'value': 'foo'
            },
            {
                'label': 'baz',
                'value': 'bar'
            }
        ]
    })
}}```

Comment (Stanton) [reply views/pulsar/v2/helpers/elem.html.twig]: I did, see this test case: 
https://github.com/jadu/pulsar/pull/906/commits/37fa82f5b627026d780f9acd56648ddd56447822#diff-0bc22e4053d041ee111cb73f1013690b

Comment (mikedevelops) [reply views/pulsar/v2/helpers/elem.html.twig]: To me it feels like we should match the documented behaviour and just perform a check on the type of the selected option (or limit the component to only accept 1 type). If somebody updated their twig version, the Pulsar unit tests wouldn't help us there would it?

Comment (Stanton) [reply views/pulsar/v2/helpers/elem.html.twig]: Sure, I see your point, happy to go this way

Comment (Stanton) [reply views/pulsar/v2/helpers/elem.html.twig]: Pushed a change which is more explicit in testing whether `selected` is a string/array.

File: tests/unit/Jadu/Pulsar/Twig/Macro/Fixtures/form/select-nested-selected-multiple-not-partial.html.twig [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+        'selected': ['ab', 'hi'],
+        'options': [
+            {
+                'label': 'abcde',
+                'value': 'abcde'
+            },
+            {
+                'label': 'fghij',
+                'value': 'fghij'
+            }
```
Comment (jackbentley) [inline tests/unit/Jadu/Pulsar/Twig/Macro/Fixtures/form/select-nested-selected-multiple-not-partial.html.twig]: Was hoping for some wittier 😞

File: views/pulsar/v2/helpers/elem.html.twig [hunk 3]
```diff
@@ -259,8 +259,18 @@ value       | string | Specifies the value of the input
                 {% for option in options %}
                     {% if option is iterable %}
                         {% for item in option %}
+
+                            {% set selectedAttr = false %}
+                            {% if mainOptions.selected is defined and mainOptions.selected is iterable %}
+                                {% if item.value in mainOptions.selected %}
```
Comment (jackbentley) [inline views/pulsar/v2/helpers/elem.html.twig]: Don't think there's any need for a nested if?

File: views/pulsar/v2/helpers/elem.html.twig [hunk 4]
```diff
@@ -275,17 +285,33 @@ value       | string | Specifies the value of the input
...
             {# non-optgroup nested syntax #}
             {% else %}
+                {% set selectedAttr = false %}
+                {% if mainOptions.selected is defined and mainOptions.selected is iterable %}
+                    {% if label.value in mainOptions.selected %}
+                        {% set selectedAttr = true %}
+                    {% endif %}
+                {% elseif mainOptions.selected is defined and mainOptions.selected == label.value %}
+                    {% set selectedAttr = true %}
+                {% endif %}
```
Comment (jackbentley) [inline views/pulsar/v2/helpers/elem.html.twig]: It's a shame this has to be copy and pasted 😞

File: views/pulsar/v2/helpers/elem.html.twig [hunk 5]
```diff
@@ -275,17 +285,33 @@ value       | string | Specifies the value of the input
...
-                    {%- if mainOptions.selected is defined and label.value == mainOptions.selected %} selected{% endif -%}
+                    {%- if selectedAttr == true %} selected{% endif -%}
                     {%- if label.disabled is defined and label.disabled == true %} disabled="disabled"{% endif -%}>
                     {{- label.label -}}
                 </option>
             {% endif %}
 
         {# legacy un-nested syntax #}
         {% else %}
+            {% set selectedAttr = false %}
```
Comment (jackbentley) [inline views/pulsar/v2/helpers/elem.html.twig]:
```suggestion
            {% set selectedAttr = false %}
            {% if mainOptions.selected is defined %}
                {% if mainOptions.selected is iterable and value in mainOptions.selected %}
                    {% set selectedAttr = true %}
                {% elseif mainOptions.selected == value %}
                    {% set selectedAttr = true %}
                {% endif %}
            {% endif %}
```

or

```suggestion
            {% set selectedAttr = false %}
            {% if mainOptions.selected is defined  %}
                {% set selectedAttr = ((mainOptions.selected is iterable and value in mainOptions.selected) or mainOptions.selected == value) %}
            {% endif %}
```

File: tests/unit/Jadu/Pulsar/Twig/Macro/Fixtures/form/select-nested-selected-multiple-not-partial.html.twig (hunk 2 above)
Comment (Stanton) [reply tests/unit/Jadu/Pulsar/Twig/Macro/Fixtures/form/select-nested-selected-multiple-not-partial.html.twig]: Sorry 🍶

**Pull request merged**